pkg/entities/record.go

```diff
@@ -101,89 +106,281 @@ func (d *dataRecord) PrepareRecord() (uint16, error) {
 	return 0, nil
 }
 
-func (d *dataRecord) AddInfoElement(element *InfoElement, val interface{}) (uint16, error) {
+func (d *dataRecord) AddInfoElement(element *InfoElementValue, isDecoding bool) (uint16, error) {
 	d.fieldCount++
 	initialLength := d.buff.Len()
+	var value interface{}
 	var err error
-	switch dataType := element.DataType; dataType {
+	if isDecoding {
+		value, err = d.decodeToIEDataType(element.Element.DataType, element.Value)
+	} else {
+		value, err = d.encodeToIEDataType(element.Element.DataType, element.Value)
+	}
+
+	if err != nil {
+		return 0, err
+	}
+	ie := NewInfoElementValue(element.Element, value)
+	d.elements = append(d.elements, ie)
+	if err != nil {
+		return 0, err
+	}
+	return uint16(d.buff.Len() - initialLength), nil
+}
+
+func (d *dataRecord) GetInfoElements() []*InfoElementValue {
+	return d.elements
+}
+
+func (t *templateRecord) PrepareRecord() (uint16, error) {
+	// Add Template Record Header
+	initialLength := t.buff.Len()
+	err := util.Encode(&t.buff, binary.BigEndian, t.templateID, t.fieldCount)
+	if err != nil {
+		return 0, fmt.Errorf("AddInfoElement(templateRecord) error in writing template header: %v", err)
+	}
+	return uint16(t.buff.Len() - initialLength), nil
+}
+
+func (t *templateRecord) AddInfoElement(element *InfoElementValue, isDecoding bool) (uint16, error) {
+	// val could be used to specify smaller length than default? For now assert it to be nil
+	if element.Value != nil {
+		return 0, fmt.Errorf("AddInfoElement(templateRecord) cannot take value %v (nil is expected)", element.Value)
+	}
+	initialLength := t.buff.Len()
+	// Add field specifier {elementID: uint16, elementLen: uint16}
+	err := util.Encode(&t.buff, binary.BigEndian, element.Element.ElementId, element.Element.Len)
+	if err != nil {
+		return 0, err
+	}
+	if element.Element.EnterpriseId != 0 {
+		// Set the MSB of elementID to 1 as per RFC7011
+		t.buff.Bytes()[initialLength] = t.buff.Bytes()[initialLength] | 0x80
+		err = util.Encode(&t.buff, binary.BigEndian, element.Element.EnterpriseId)
+		if err != nil {
+			return 0, err
+		}
+	}
+	t.elements = append(t.elements, element)
+	// Keep track of minimum data record length required for sanity check
+	if element.Element.Len == VariableLength {
+		t.minDataRecLength = t.minDataRecLength + 1
+	} else {
+		t.minDataRecLength = t.minDataRecLength + element.Element.Len
+	}
+	return uint16(t.buff.Len() - initialLength), nil
+}
+
+func (t *templateRecord) GetInfoElements() []*InfoElementValue {
+	return t.elements
+}
+
+func (t *templateRecord) GetMinDataRecordLen() uint16 {
+	return t.minDataRecLength
+}
+
+// decodeToIEDataType is to decode to specific type
+func (d *dataRecord) decodeToIEDataType(dataType IEDataType, val interface{}) (interface{}, error) {
```

Comment:
Does it need to be a method of dataRecord? You could define it as utility function.

Reply:
Makes sense. I moved the functions to `ie.go` since they are info element related. Thanks!